### PR TITLE
Less zip false positives

### PR DESCRIPTION
**Pull request by lhoestq**
`zipfile.is_zipfile` return false positives for some Parquet files. It causes errors when loading certain parquet datasets, where some files are considered ZIP files by `zipfile.is_zipfile`

This is a known issue: https://github.com/python/cpython/issues/72680

At first I wanted to rely only on magic numbers, but then I found that someone contributed a [fix to is_zipfile](https://github.com/python/cpython/pull/5053) - do you think we should use it @albertvillanova or not ?

IMO it's ok to rely on magic numbers only for now, since in streaming mode we've had no issue checking only the magic number so far.

Close https://github.com/huggingface/datasets/issues/5639

### CODE REVIEW

**Comment by lhoestq** [comment]
CI is failing due to unrelated issues, hopefilly https://github.com/huggingface/datasets/pull/5642 fixes it

File: src/datasets/utils/extract.py [hunk 1]
```diff
@@ -137,10 +139,48 @@ def extract(input_path: Union[Path, str], output_path: Union[Path, str]) -> None
...
+                    elif endrec[_ECD_DISK_NUMBER] == endrec[_ECD_DISK_START]:
+                        fp.seek(endrec[_ECD_OFFSET])  # Central directory is on the same disk
+                        if fp.tell() == endrec[_ECD_OFFSET] and endrec[_ECD_SIZE] >= sizeCentralDir:
+                            data = fp.read(sizeCentralDir)  # CD is where we expect it to be
+                            if len(data) == sizeCentralDir:
+                                centdir = struct.unpack(structCentralDir, data)  # CD is the right size
+                                if centdir[_CD_SIGNATURE] == stringCentralDir:
+                                    return True  # First central directory entry  has correct magic number
+        except Exception:  # catch all errors in case future python versions change the zipfile internals
+            return False
```
Comment (albertvillanova) [inline src/datasets/utils/extract.py:183]: Please note that this function as it is could return `None`. To fix this:
```suggestion
            return False
        return False
```

File: tests/test_extract.py [hunk 2]
```diff
@@ -183,3 +184,20 @@ def test_tar_extract_insecure_files(
...
+    data = (
+        b"\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x00\x01\x00\x00"
+        b"\x00\x02\x08\x06\x00\x00\x00\x99\x81\xb6'\x00\x00\x00\x15I"
+        b"DATx\x01\x01\n\x00\xf5\xff\x00PK\x05\x06\x00PK\x06\x06\x07"
+        b"\xac\x01N\xc6|a\r\x00\x00\x00\x00IEND\xaeB`\x82"
+    )
+    with not_a_zip_file.open("wb") as f:
+        f.write(data)
+    assert zipfile.is_zipfile(str(not_a_zip_file))  # is a false positive for `zipfile`
+    assert not ZipExtractor.is_extractable(not_a_zip_file)  # but we're right
```
Comment (albertvillanova) [inline tests/test_extract.py:203]: The test passes because `not None` is `True`.

**Review by albertvillanova — APPROVED** [review]
Thanks for the improvement.

I agree we should tweak the `zipfile.is_zipfile` default implementation if that is flaky.

File: src/datasets/utils/extract.py [hunk 3]
```diff
@@ -137,10 +139,48 @@ def extract(input_path: Union[Path, str], output_path: Union[Path, str]) -> None
...
+                    if endrec[_ECD_ENTRIES_TOTAL] == 0 and endrec[_ECD_SIZE] == 0 and endrec[_ECD_OFFSET] == 0:
+                        return True  # Empty zipfiles are still zipfiles
+                    elif endrec[_ECD_DISK_NUMBER] == endrec[_ECD_DISK_START]:
+                        fp.seek(endrec[_ECD_OFFSET])  # Central directory is on the same disk
+                        if fp.tell() == endrec[_ECD_OFFSET] and endrec[_ECD_SIZE] >= sizeCentralDir:
+                            data = fp.read(sizeCentralDir)  # CD is where we expect it to be
+                            if len(data) == sizeCentralDir:
+                                centdir = struct.unpack(structCentralDir, data)  # CD is the right size
+                                if centdir[_CD_SIGNATURE] == stringCentralDir:
+                                    return True  # First central directory entry  has correct magic number
```
Comment (albertvillanova) [inline src/datasets/utils/extract.py:181]: The fix could also be added here:
```suggestion
                                    return True  # First central directory entry  has correct magic number
            return False
```

**Review by albertvillanova — APPROVED** [review]

**Pull request merged**